typeurl/types.go

```diff
@@ -0,0 +1,99 @@
+package typeurl
+
+import (
+	"encoding/json"
+	"errors"
+	"path/filepath"
+	"reflect"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+)
+
+const Namespace = "types.containerd.io"
+
+var (
+	mu            sync.Mutex
+	registry      = make(map[reflect.Type]string)
+	ErrRegistered = errors.New("typeurl: type already registred")
+	ErrNotExists  = errors.New("typeurl: type is not registered")
+)
+
+// Register a type with the base url of the type
+func Register(v interface{}, args ...string) {
+	t := tryDereference(v)
+	mu.Lock()
+	defer mu.Unlock()
+	if _, ok := registry[t]; ok {
+		panic(ErrRegistered)
+	}
+	registry[t] = filepath.Join(append([]string{Namespace}, args...)...)
+}
+
+// TypeUrl returns the type url for a registred type
+func TypeUrl(v interface{}) (string, error) {
+	mu.Lock()
+	defer mu.Unlock()
+	u, ok := registry[tryDereference(v)]
+	if !ok {
+		return "", ErrNotExists
```

Comment:
See `events/convert.go` and `proto.MessageName(pb)`. I think you can dispatch on `proto.Message`.